lib/src/restoration.dart

```diff
@@ -0,0 +1,241 @@
+    super.didUpdateHook(oldHook);
+    if (hook.scope != oldHook.scope ||
+        hook.restorationId != oldHook.restorationId) {
+      oldHook.scope.unRegisterProperty(oldHook.restorationId);
+      hook.scope.registerProperty(hook.restorationId, hook.property);
+    }
+  }
+
+  @override
+  bool shouldRebuild() => false;
+
+  @override
+  void dispose() {
+    super.dispose();
+    property.dispose();
+  }
+}
+
+/// Registers [property] with the nearest [HookRestorationScope] under
+/// [restorationId].
+///
+/// Only the value passed for [property] when [useRestorationProperty] is called
+/// for the first time is used. Every time [useRestorationProperty] is called
+/// the same initial value is returned.
+///
+/// By default the registered property is listened to and triggers a rebuild of
+/// the surrounding widget when changed. Setting [listen] to `false` disables
+/// this behaviour.
+///
+/// ## Example
+///
+/// ```dart
+/// class RestorablePropertyExample extends StatelessWidget {
+///   @override
+///   Widget build(BuildContext context) {
+///     return Scaffold(
+///       appBar: AppBar(
+///         title: const Text('useRestorationProperty example'),
+///       ),
+///       body: RootRestorationScope(
+///         restorationId: 'root',
+///         child: HookRestorationScope(
+///           restorationId: 'hooks',
+///           child: HookBuilder(
+///             builder: (context) {
+///               final count = useRestorationProperty('count', RestorableInt(0));
+///               return Center(
+///                 child: ElevatedButton(
+///                   onPressed: () => count.value += 1,
+///                   child: Text('${count.value} clicks'),
+///                 ),
+///               );
+///             },
+///           ),
+///         ),
+///       ),
+///     );
+///   }
+/// }
+/// ```
+///
+/// See also:
+///
+///   * [HookRestorationScope] which is equivalent to [RestorationScope] but
+///     additionally enables the usage of this hook in its children.
+///   * [RestorableProperty] which is a wrapper for a value which should be
+///     persisted and restored as part of Flutters restoration system.
+T useRestorationProperty<T extends RestorableProperty<void>>(
+  String restorationId,
+  T property, {
+  bool listen = true,
+}) {
+  assert(restorationId != null, 'restorationId cannot be null.');
+  assert(property != null, 'property cannot be null.');
+  assert(listen != null, 'listenToValue cannot be null.');
+
+  final scope = useContext()
+      .dependOnInheritedWidgetOfExactType<_HookRestorationScopeMarker>()
+      ?.scope;
+
+  assert(
+    scope != null,
+    'To use useRestorationProperty make sure a HookRestorationScope is '
+    'available from where useRestorationProperty is called.',
+  );
+
+  final result = use(_RestorablePropertyHook(
+    restorationId: restorationId,
+    property: property,
+    scope: scope,
+  ));
+
+  if (listen) {
+    useListenable(result);
```

Comment:
Conditionally calling hooks is anti-pattern. The `listen` flag probably shouldn't be there at all

Reply:
I assumed most users will want to listen to the property changes but thinking about it I'm not so sure. I'll remove the option and document that properties need to be listened to explicitly for example with `useListenable`.

Reply:
Sorry, I meant "always listen" instead, rather than optionally listen.

If we want to support optional listening, this should probably be in a different hook rather than as a flag